src/passes/StackIR.cpp

```diff
@@ -350,6 +353,115 @@ class StackIROptimizer {
     // Otherwise, for basic instructions, just count the expression children.
     return ChildIterator(inst->origin).children.size();
   }
+
+  // Given a pair of a local.set and local.get, see if we can remove them
+  // without breaking validation. Specifically, we must keep sets of non-
+  // nullable locals that dominate a get until the end of the block, such as
+  // here:
+  //
+  //  local.set 0    ;; Can we remove
+  //  local.get 0    ;; this pair?
+  //  if
+  //    local.set 0
+  //  else
+  //    local.set 0
+  //  end
+  //  local.get 0    ;; This get poses a problem.
+  //
+  // Logically the 2nd&3rd sets ensure a value is applied to the local before we
+  // read it, but the validation rules only track each set until the end of its
+  // scope, so the 1st set (before the if, in the pair) is necessary.
+  //
+  // The logic below is related to LocalStructuralDominance, but sharing code
+  // with it is difficult as this uses StackIR and not BinaryenIR, and it checks
+  // a particular set/get pair.
+  //
+  // We are given the indexes of the set and get instructions in |insts|.
+  bool canRemoveSetGetPair(Index setIndex, Index getIndex) {
+    // The set must be before the get.
+    assert(setIndex < getIndex);
+
+    auto* set = insts[setIndex]->origin->cast<LocalSet>();
+    if (func->isParam(set->index) ||
+        !func->getLocalType(set->index).isNonNullable()) {
+      // This local cannot pose a problem for validation (params are always
+      // initialized, and nullable locals may be uninitialized).
+      return true;
+    }
+
+    // Track the depth (in block/if/loop/etc. scopes) relative to our starting
+    // point. Anything less deep than that is not interesting, as we can only
+    // help things at our depth or deeper to validate.
+    Index currDepth = 0;
+
+    // Look for another get than the one in getIndex (since that one is being
+    // removed) which would stop validating without us. While doing so, note
+    // other sets that ensure validation even if our set is removed. We track
+    // those in this stack of booleans, one for each scope, which is true if
+    // another sets covers us and ours is not needed.
+    //
+    // We begin in the current scope and with no other set covering us.
+    std::vector<bool> coverStack = {false};
+
+    // Track the total number of covers as well, for quick checking below.
+    Index covers = 0;
+
+    // TODO: We could look before us as well, but then we might end up scanning
+    //       much of the function every time.
```

Comment:
Good idea. That would be guaranteed linear time which sounds nice. I'll look into that tomorrow.

Reply:
I guess it's technically quadratic because each set could contribute to the counter for multiple gets, but it would probably be ok in practice.

Reply:
Perhaps another approach rather than using counters would be to build a structural dominance tree including sets dominated by other sets with gets as the leaves. Removing sets would remove their nodes from the tree and would be blocked if a get would have no more parent. I think that solution actually would be linear, but it would also be more complicated.

Reply:
Actually, we really do need a dominance tree or some other way to quickly find all the gets dominated by each set. Otherwise, each time we try to remove a set we need to scan for its gets. I've written most of the code for that idea but it's already more complicated than this PR, and I'm not certain it's actually going to be more efficient. So I think this PR as is is better: in the general case we don't need to generate all this info, and can just do a linear scan for each set we try to remove.

The problem of sets before us enabling validation remains, so we'll miss some opportunities, but getting 80% of this minor optimization seems good enough.